parsers/JP.py

```diff
@@ -352,9 +361,178 @@ def parse_dt(row):
+        "JP-HR": f"https://www.rikuden.co.jp/nw/denki-yoho/csv/yosoku_05_{datestamp}.csv",
+        "JP-KN": f"https://www.kansai-td.co.jp/interchange/denkiyoho/imbalance/{datestamp}_yosoku.csv",
+        "JP-CG": f"https://www.energia.co.jp/nw/jukyuu/sys/{datestamp[:6]}_jyukyu2_chugoku.zip",
+        "JP-KY": rf"https://www.kyuden.co.jp/td_power_usages/csv/kouhyo/imbalance/21110_TSO9_0_{datestamp}.csv?a={datestamp}\d{6}",
+        "JP-ON": f"http://www.okiden.co.jp/denki2/dem_pg/csv/jukyu_yosoku_{datestamp}.csv",
+    }
+
+    # For zones with different URLs for today vs tomorrow
+    if datestamp == datetime.now(ZONE_INFO).strftime("%Y%m%d"):
+        forecast_url["JP-TH"] = (
+            f"https://setsuden.nw.tohoku-epco.co.jp/common/demand/area_tso_yosoku_{datestamp}.csv"
+        )
+        forecast_url["JP-TK"] = (
+            "https://www4.tepco.co.jp/forecast/html/images/AREA_YOSOKU.csv"
+        )
+        forecast_url["JP-CB"] = (
+            "https://powergrid.chuden.co.jp/denki_yoho_content_data/keito_yosoku_cepco003.csv"
+        )
+        forecast_url["JP-SK"] = (
+            "https://www.yonden.co.jp/nw/denkiyoho/supply_demand/csv/yosoku_today.csv"
+        )
+    # TODO: what time the new doc is created for next day?
+    elif datestamp == (datetime.now(ZONE_INFO) + timedelta(days=1)).strftime("%Y%m%d"):
+        forecast_url["JP-TH"] = (
+            "https://setsuden.nw.tohoku-epco.co.jp/common/demand/area_tso_yosoku_y.csv"
+        )
+        forecast_url["JP-TK"] = (
+            "https://www4.tepco.co.jp/forecast/html/images/AREA_ONCE_YOSOKU.csv"
+        )
+        forecast_url["JP-SK"] = (
+            "https://www.yonden.co.jp/nw/denkiyoho/supply_demand/csv/yosoku_tomorrow.csv"
+        )
+        if zone_key == "JP-CB":  # Raise exception for JP-CB
+            raise ValueError(
+                f"Forecast data for tomorrow not available for zone: {zone_key}"
+            )
+
+    # Skip non-tabular data at the start of source files
+    startrow = 1 if zone_key == "JP-HR" else 3 if zone_key == "JP-KN" else 2
+
+    # Get response from correspondent url
+    response = session.get(forecast_url[zone_key])
+
+    # Parse the different formats
+    if zone_key == "JP-CG":
+        with ZipFile(BytesIO(response.content)) as z:
+            target_file = f"{datestamp}_yosoku_chugoku.csv"
+            if target_file in z.namelist():
+                with z.open(target_file) as f:
+                    df = pd.read_csv(f, encoding="shift_jis", skiprows=startrow)
+            else:
+                print(f"File '{target_file}' not found in the zip archive.")
+                print(f"Available files: {z.namelist()}")
+    else:
+        content = response.content.decode("shift_jis")
+        df = pd.read_csv(StringIO(content), skiprows=startrow)
+
+    # Extract the correspondant information from the files for each zone
+    if zone_key == "JP-HKD" or zone_key == "JP-HR":
+        df = df[["日付", "時間帯_自", "エリア総発電量(kWh)"]]
```

Comment:
Since the df is created in a conditional above it's potentially unbound here and downstream.